### PR TITLE
Fix record-by-defauilt

### PR DESCRIPTION
**Pull request by apotterri**
Make sure it instruments application code. There's definitely more testing that could be added here, but [this](https://github.com/getappmap/appmap-python/pull/180/commits/ce0b5239b13608db2874bed175d193267af4ce82#diff-f0da5d71b031834061fcaa37daf4a32c38ab2e6030993abdcbb8e0d7036c22f5R112) is a start.

Fixes getappmap/board#191.

### CODE REVIEW

File: appmap/test/web_framework.py [hunk 1]
```diff
@@ -363,6 +369,10 @@ def server_url():
 
     @staticmethod
     def record_request_thread():
+        # I've seen occasional test failures, seemingly because the test servers can't handle the
+        # barrage of requests. A tiny bit of delay still causes many, many concurrent requests, but
+        # eliminates the failures.
+        time.sleep(SR.uniform(0, 0.1))
```
Comment (symwell) [inline appmap/test/web_framework.py:375]: I ran `poetry run pytest -s -v appmap/test/test_flask.py` while also running saleor tests and got:
```
            # wait for all threads to complete
            for future in concurrent.futures.as_completed(future_to_request_number):
                try:
                    response = future.result()
                except Exception as e:
                    traceback.print_exc()
>               assert response.status_code == 200
E               assert 400 == 200
E                +  where 400 = <Response [400]>.status_code

appmap/test/web_framework.py:410: AssertionError
```

Also ran when not running anything else and got:

```
FAILED appmap/test/test_flask.py::TestRecordRequestsFlask::test_record_request_appmap_enabled_requests_enabled_no_remote - assert 400 == 200
FAILED appmap/test/test_flask.py::TestRecordRequestsFlask::test_record_request_appmap_enabled_requests_enabled_and_remote - assert 400 == 200
FAILED appmap/test/test_flask.py::TestRecordRequestsFlask::test_record_request_appmap_not_enabled_requests_enabled_no_remote - assert 400 == 200
```

Let's keep this change though.  I can imagine it helps. The real problem may be the ability of the tested webserver (Flask/Django) to handle concurrency.

File: appmap/flask.py [hunk 2]
```diff
@@ -113,16 +117,15 @@ def before_request(self):
...
-            np = "".join(
-                [
-                    f"{{{p}}}" if c else p
-                    for c, _, p in parse_rule(request.url_rule.rule)
-                ]
+            # Transform requestl.url to the expected normalized-path form. For example,
+            # "/post/<username>/<post_id>/summary" becomes "/post/{{username}}/{{post_id}}/summary".
+            r = repr(request.url_rule)
+            np = NP_XLATE.sub(
+                lambda m: "{" if m["l"] else "}" if m["r"] else "",
```
Comment (dividedmind) [inline appmap/flask.py:125]: Should this be double `{{` and `}}` or is the comment inaccurate? Regardless, the comment could show the `repr` format so it's clear what's happening.

(I'm kind of weirded out by using `repr` like this, but as long as it works... also, let's hope rules never have a `'` in the name.)

File: appmap/flask.py [hunk 3]
```diff
@@ -113,16 +117,15 @@ def before_request(self):
...
+                NP_REGEXP.findall(r)[0],
```
Comment (dividedmind) [inline appmap/flask.py:126]: I don't think this works. I tried `r = '/post/<username>/<post_id>/summary'` — I got `IndexError: list index out of range`. I tried `r = '/post/<Rule \'username\'>/<Rule \'post_id\'>/summary'` — I got `np = 'username'`. Do we have tests for this?

Also, does this even belong in this pull request?

File: appmap/test/test_configuration.py [hunk 4]
```diff
@@ -47,19 +45,24 @@ def test_reports_invalid():
...
+@pytest.mark.appmap_enabled(config="appmap-broken.yml", appmap_enabled="false")
 def test_is_disabled_when_false():
     """Test that recording is disabled when APPMAP=false"""
     Env.current.set("APPMAP", "false")
     assert not appmap.enabled()
 
 
+@pytest.mark.skip(
+    reason="TODO: determine if this test makes sense, appmap is enabled by default"
+)
```
Comment (dividedmind) [inline appmap/test/test_configuration.py:65]: Yeah, should probably just remove it.

**Review by dividedmind — CHANGES_REQUESTED** [review]
The detect enabled part seems fine. But there seems to be a stray broken flask commit there.

**Comment by symwell** [comment]
Is it worth adding a testcase that expects the importer to work `if Env.current.enabled`?

File: appmap/flask.py (hunk 2 above)
Comment (apotterri) [reply appmap/flask.py:125]: The comment is wrong. I'll add the `repr` format.

Using `repr` feels better to me than reaching into the implementation and using an undocumented interface (which was completely rewritten for Flask 2). The `repr` format is the same for both versions, so it should work for a least a little while....

File: appmap/flask.py (hunk 3 above)
Comment (apotterri) [reply appmap/flask.py:126]: Yes, there's a test for it, but I'll back and look at what's going on with.

Well, including it seemed harmless, but clearly it's not working well enough to keep. I'll take it out, and fix whatever problems it has as part of the Flask 2 work.

File: appmap/test/test_configuration.py (hunk 4 above)
Comment (apotterri) [reply appmap/test/test_configuration.py:65]: WIll do.

**Comment by apotterri** [comment]
> Is it worth adding a testcase that expects importer.py to work if Env.current.enabled?

If you're asking for a test case that confirms that application code gets instrumented for record-by-default, that's effectively what deleting `APPMAP` from the environment does here: https://github.com/getappmap/appmap-python/blob/ce0b5239b13608db2874bed175d193267af4ce82/appmap/test/test_test_frameworks.py#L112 (this was the change the link in my original comment was supposed to point to, sorry about that). I went back and confirmed that if I made that change to the code on `master`, test failed because application code didn't appear in the generated recordings.

If that's not what you're asking for, though, maybe you could clarify a bit? Thanks!

File: appmap/test/web_framework.py (hunk 1 above)
Comment (symwell) [reply appmap/test/web_framework.py:375]: The failures I saw turned out to be due to a saleor container using the same port number, 8000.
All tests pass after I stopped this container.  All good.

**Comment by symwell** [comment]
> If that's not what you're asking for, though, maybe you could clarify a bit?

I might not fully understand how appmap gets enabled.

I thought appmap gets enabled in two ways
- from the code in `appmap/__init__.py` installed through pip.  This part works.  When a standalone Python program in a single file runs, all the functions in it are instrumented.
- when `import some_module_name` is used by a program.  I thought something more needs to happen for appmap to instrument code for this import.  I thought there would be a testcase in two standalone files where the first tries to import the second.

File: appmap/flask.py [hunk 5]
```diff
@@ -57,7 +57,7 @@ def init_app(self, app):
             # it may record requests but not remote (APPMAP=false)
             self.recorder = Recorder.get_current()
 
-        if Env.current.enabled:
+        if DetectEnabled.should_enable("remote"):
             # the remote recording routes are enabled only if APPMAP=true
```
Comment (dividedmind) [inline appmap/flask.py:61]: The comment is stale now.

**Review by dividedmind — APPROVED** [review]
Looks good, thanks :)

File: appmap/flask.py (hunk 5 above)
Comment (apotterri) [reply appmap/flask.py]: Thanks, fixed.

**Pull request merged**